packages/ckeditor5-source-editing/src/sourceediting.js

```diff
@@ -335,3 +335,105 @@ export default class SourceEditing extends Plugin {
 		return editable && !editable._hasExternalElement;
 	}
 }
+
+// Beautifies the input HTML string by adding new lines and indenting some HTML elements.
+//
+// @param {String} input HTML string to beautify.
+// @returns {String}
+function formatSource( input ) {
+	const elementsToFormat = [
+		{ name: 'blockquote', isVoid: false },
+		{ name: 'br', isVoid: true },
+		{ name: 'div', isVoid: false },
+		{ name: 'figcaption', isVoid: false },
+		{ name: 'figure', isVoid: false },
+		{ name: 'h1', isVoid: false },
+		{ name: 'h2', isVoid: false },
+		{ name: 'h3', isVoid: false },
+		{ name: 'h4', isVoid: false },
+		{ name: 'h5', isVoid: false },
+		{ name: 'h6', isVoid: false },
+		{ name: 'hr', isVoid: true },
+		{ name: 'li', isVoid: false },
+		{ name: 'ol', isVoid: false },
+		{ name: 'p', isVoid: false },
+		{ name: 'table', isVoid: false },
+		{ name: 'tbody', isVoid: false },
+		{ name: 'td', isVoid: false },
+		{ name: 'th', isVoid: false },
+		{ name: 'thead', isVoid: false },
+		{ name: 'tr', isVoid: false },
+		{ name: 'ul', isVoid: false }
```

Comment:
Defining an array of all elements that should be formatted is quite risky to me. Note that e.g. `<img>` is missing here. And this array should be actively maintained in the future. 

I'd rather collect only all possible tags that are empty and handle the rest tags in a more generic way.

Reply:
Maybe we could extend it with at least all defined block elements - https://developer.mozilla.org/en-US/docs/Web/HTML/Block-level_elements.

Reply:
`isVoid` -> `isEmpty` ? I guess this is the proper naming for these elements.

Reply:
I will add all block elements to this list.

> `isVoid` -> `isEmpty` ? I guess this is the proper naming for these elements.

According to W3C naming convention such elements are "void": https://html.spec.whatwg.org/multipage/syntax.html#void-elements. The term "empty element" is used to describe an element that contains no content. Elements of any type can be empty, depending only on their content model (i.e. `<span></span>` is an empty element). So IMHO we should keep the `isVoid` property as it is now.

Reply:
I took this name from [https://developer.mozilla.org/en-US/docs/Glossary/Empty_element.](https://developer.mozilla.org/en-US/docs/Glossary/Empty_element.) But we can use void if this is a good name for these elements. My main concern here is rather https://github.com/ckeditor/ckeditor5/pull/9875#discussion_r650951288.

Reply:
One more option to consider: `isSelfClosing`.

Reply:
A self-closing tag ends with `/>` which is not true for void elements. Our editor produces `<br>` but not `<br />`.

Reply:
> Maybe we could extend it with at least all defined block elements - https://developer.mozilla.org/en-US/docs/Web/HTML/Block-level_elements.

Because we shouldn't format all elements, we have to define somehow list of elements, that will be formatted. I don't know how it could have been done otherwise than introducing an array of them in the code.